src/platform/ESP32/nimble/BLEManagerImpl.cpp

```diff
@@ -0,0 +1,947 @@
+{
+    CHIP_ERROR err;
+
+    mServiceMode = ConnectivityManager::kCHIPoBLEServiceMode_Enabled;
+    /*XXX for (int i = 0; i < kMaxConnections; i++)
+    {
+
+
+    }*/
+
+    // Initialize the Chip BleLayer.
+    err = BleLayer::Init(this, this, &SystemLayer);
+    SuccessOrExit(err);
+
+    mServiceAttrHandle    = 0;
+    mRXCharAttrHandle     = 0;
+    mTXCharAttrHandle     = 0;
+    mTXCharCCCDAttrHandle = 0;
+    mFlags       = kFlag_AdvertisingEnabled;
+    mNumGAPCons  = 0;
+    memset(mCons, 0, sizeof(mCons));
+    mServiceMode          = ConnectivityManager::kCHIPoBLEServiceMode_Enabled;
+    mFlags                = kFlag_AdvertisingEnabled;
+    memset(mDeviceName, 0, sizeof(mDeviceName));
+
+    ble_svc_gap_init();
+    ble_svc_gatt_init();
+
+    err = ble_gatts_count_cfg(CHIPoBLEGATTAttrs);
+    SuccessOrExit(err);
+    err = ble_gatts_add_svcs(CHIPoBLEGATTAttrs);
+    SuccessOrExit(err);
+
+
+    PlatformMgr().ScheduleWork(DriveBLEState, 0);
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BLEManagerImpl::_SetCHIPoBLEServiceMode(CHIPoBLEServiceMode val)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(val != ConnectivityManager::kCHIPoBLEServiceMode_NotSupported, err = CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrExit(mServiceMode == ConnectivityManager::kCHIPoBLEServiceMode_NotSupported, err = CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
+
+    if (val != mServiceMode)
+    {
+        mServiceMode = val;
+        PlatformMgr().ScheduleWork(DriveBLEState, 0);
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BLEManagerImpl::_SetAdvertisingEnabled(bool val)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    VerifyOrExit(mServiceMode == ConnectivityManager::kCHIPoBLEServiceMode_NotSupported, err = CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE);
+
+    if (GetFlag(mFlags, kFlag_AdvertisingEnabled) != val)
+    {
+        SetFlag(mFlags, kFlag_AdvertisingEnabled, val);
+        PlatformMgr().ScheduleWork(DriveBLEState, 0);
+    }
+
+exit:
+    return err;
+}
+
+CHIP_ERROR BLEManagerImpl::_GetDeviceName(char * buf, size_t bufSize)
+{
+    if (strlen(mDeviceName) >= bufSize)
+    {
+        return CHIP_ERROR_BUFFER_TOO_SMALL;
+    }
+    strcpy(buf, mDeviceName);
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BLEManagerImpl::_SetDeviceName(const char * deviceName)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+
+    if (mServiceMode == ConnectivityManager::kCHIPoBLEServiceMode_NotSupported)
+    {
+        return CHIP_ERROR_UNSUPPORTED_CHIP_FEATURE;
+    }
+    if (deviceName != NULL && deviceName[0] != 0)
+    {
+        if (strlen(deviceName) >= kMaxDeviceNameLength)
+        {
+            return CHIP_ERROR_INVALID_ARGUMENT;
+        }
+        strcpy(mDeviceName, deviceName);
+        err = ble_svc_gap_device_name_set(mDeviceName);
+        SuccessOrExit(err);
+    }
+    else
+    {
+        mDeviceName[0] = 0;
+        ClearFlag(mFlags, kFlag_UseCustomDeviceName);
+    }
+    return CHIP_NO_ERROR;
+
+exit:
+    return CHIP_ERROR_INCORRECT_STATE;
```

Comment:
Should we preserve 'err' and have

```
exit:
    return err;
```

instead? here we lose the err from the call on line 225 and we get a generic 'incorrect state' which loses some error info.

Reply:
Have replaced with `err`.